File: Lab/ME4/fork_and_exec.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>

void initiateFork() {

	pid_t pid;
	int returnValue;

	switch(pid = fork()) {
		case -1: 
				perror("Fork error.\n");
				exit(1);

		case 0: 
				printf("CHILD\n");
				printf("\tPID IS: %d\n", getpid());
				printf("\tPARENT PID IS: %d\n", getppid());
				printf("\tSetting status exit to 44...\n");
				returnValue = 44;
				printf("\tCHILD: Time to rest.\n");
				exit(returnValue);

		default: 
				printf("PARENT\n");
				printf("\tPID IS: %d\n", getpid());
				printf("\tPARENT PID IS: %d\n", getppid());
				printf("\tWaiting for child to terminate...\n");
				wait(&returnValue);
				printf("\tChild exited with status: %d\n", WEXITSTATUS(returnValue));
				printf("\tPARENT: Time to rest.\n");
	}
}

void initiateExec() {
	printf("\nCalling 'cowsay' with execvp()...\n");
	char *args[]={"cowsay", "\"KIM CHAEWON BEST GIRL\"" ,NULL}; 
	execvp(args[0], args); 
}

int main() {

	initiateFork();
	initiateExec();

	return 0;
}

File: Lab/ME4/fork_and_exec.py
import os
import sys

def initiateFork():
	print("Initiating fork...\n")
	pid = os.fork()

	if(pid == 0): 
		print("CHILD") 
		print("\tPID is:", os.getpid()) 
		print("\tPARENT PID IS:", os.getppid()) 
		print("\tCHILD: Time to rest.");
		print("\nFork done!")

	else: 
		print("PARENT") 
		print("\tPID is:", os.getpid()) 
		print("\tPARENT PID IS:", os.getppid()) 
		print("\tWaiting for child to terminate...")
		print("\tPARENT: Time to rest.\n")


def initiateExec():
	print("Initiating exec...")
	program = """print(\"\"\"
 _______________________
< KIM CHAEWON BEST GIRL >
 -----------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                
                ||     ||

		\"\"\")"""
	exec(program)
	print("Exec done!\n")

if __name__ == '__main__':
	initiateExec()
	initiateFork()